rust/exchange_rate/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+        "Via",
+        "X-Amz-Cf-Id",
+    ];
+}
+
+/// Canister heartbeat. Process one item in queue
+// #[export_name = "canister_heartbeat"]
+#[heartbeat]
+async fn heartbeat() {
+    ic_cdk::println!("Heartbeat method invoked.");
+    get_next_rate().await;
+}
+
+/*
+Get rates for a time range defined by start time and end time. This function can be invoked
+as HTTP update call.
+*/
+#[update]
+#[candid_method(update)]
+async fn get_rates(start: Timestamp, end: Timestamp) -> HashMap<Timestamp, Rate> {
+    // round down start time and end time to the minute (chop off seconds), to be checked in the hashmap
+    let start_min = start / 60;
+    let end_min = end / 60;
+
+    // compose a return structure
+    let mut fetched = HashMap::new();
+
+    // pull available ranges from hashmap
+    FETCHED.with(|map_lock| {
+        let map = map_lock.borrow();
+        for requested_min in start_min..end_min {
+            let requested = requested_min * 60;
+            if map.contains_key(&requested) {
+                // The fetched slot is within user requested range. Add to result for later returning.
+                fetched.insert(requested, map.get(&requested).unwrap().clone());
+            } else {
+                // asynchoronously request downloads for unavailable ranges
+
+                // Simply putting the request to request queue. This queue will likely
+                // have duplicate entries, if users request same range of data multiple times.
+                // Double downloading is avoided right before the time of downloading by checking
+                // whether the data already exists in FETCHED map.
+                REQUESTED.with(|requested_lock| {
+                    let mut queue = requested_lock.borrow_mut();
+                    match queue.add(requested) {
+                        Ok(_) => {}
+                        Err(failure) => {
+                            ic_cdk::println!(
+                                "Wasn't able to add job {} to queue. Receiving error {}",
+                                requested,
+                                failure
+                            );
+                        }
+                    }
+                });
+            }
+        }
+    });
+
+    // return rates for available ranges
+    return fetched;
+}
+
+/*
+Register the cron job which take the tip of the queue, and send a canister call to self.
+Potentially, different nodes executing the canister will trigger different job during the same period.
+The idea is to gap the cron job with large enough time gap, so they won't trigger remove service side
+rate limiting.
+ */
+async fn get_next_rate() -> () {
+    let mut job_id: u64 = 0;
+
+    // Get the next downloading job
+    REQUESTED.with(|requested_lock| {
+        let mut requested = requested_lock.borrow_mut();
+
+        if requested.size() == 0 {
+            return;
+        }
+
+        let job = requested.remove();
+
+        match job {
```

Comment:
The only error returned by remove() is "queue is empty", so you can just do:

match requested_lock.borrow_mut().remove() { 
...
  Err(_) => return

  async fn get_next_rate() -> () {
      let mut job_id: u64 = 0;
  
      // Get the next downloading job
      REQUESTED.with(|requested_lock| {
          match requested_lock.borrow_mut().remove() {
              Ok(job) => {
                  // Job is a valid Job Id
                  FETCHED.with(|fetched_lock| {
                      match fetched_lock.borrow().get(&job) {
                          Some(_) => {
                              // If this job has already been downloaded. Only downloading it if doesn't already exist.
                              ic_cdk::println!(
                                  "Rate for {} is already downloaded. Skipping downloading again.",
                                  job
                              );
                              return;
                          }
                          None => {
                              // The requested time rate isn't found in map. Send a canister get_rate call to self
                              ic_cdk::println!("Fetching job {job} now.");
                              job_id = job;
                          }
                      }
                  });
              }
              Err(_) => return,
          }
      });
      get_rate(job_id).await;
  }